### DISCUSSION TITLE
Nextjs dynamic to lazy load components for SEO

### DISCUSSION THREAD

**Discussion by nonoumasy** [vercel/next.js#43502] [Help]
Just wanted to confirm something.
I'm using nextjs dynamic https://nextjs.org/docs/advanced-features/dynamic-import to lazy load components so that:
1. things load faster
2. javascript files are split into smaller files (for better SEO performance).

Is this right?

**Comment by nonoumasy** [comment]
anyone...

**Comment by icyJoseph** [comment]
This is not necessarily only SEO related.

The general concept under the hood is code-splitting. https://reactjs.org/docs/code-splitting.html.

> React itself exposes `lazy`, though `next/dynamic`, let's say, is a tweaked version of it. 

Shipping a very large JS bundle will cause large download, parsing and execution time. If we code-split to load only the things we need we cut down on all three dimensions. There's also rendering but I think it is not impacted as much by code-splitting.

Also, though the main thread is freed earlier, the dynamically loaded module still has to be downloaded, parsed and executed, so it is not a magic process by any means, it'll just happen later, hopefully without interrupting the user.

At it is heart this is a form of loading JS modules dynamically.

There's of course gotchas!

- During SSR, Next.js renders the dynamically loaded component, [but it can also skip it](https://nextjs.org/docs/advanced-features/dynamic-import#with-no-ssr)
- Code chunks have a bit of overhead, so the cost of code-splitting a small component, might be higher than its benefit.
- You can use a non-browser client, such as [VSCode Rest Client](https://marketplace.visualstudio.com/items?itemName=humao.rest-client] or Postman, or just turn off JS on your browser, to see what a very basic SEO bot would see
- Code splitting might cause layout shifts accumulation, as the bundles and loading spinners are replaced by content. This is bad for SEO.
- Often you'd end up using a library not ready for SSR, so you put it behind the first gotcha, but now the HTML for that section is not part of the initial HTML frame.
- You might leak sensitive data, and chunks are not tree-shaken by default, you'd need to start to control Webpack's API via magic omments: `/* webpackExports: ["module-you-want-to-load"] */`
